pkg/user/tx_client.go

```diff
@@ -166,83 +170,124 @@ func SetupTxClient(
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	if err := s.checkAccountLoaded(ctx, account); err != nil {
+func (client *TxClient) BroadcastPayForBlobWithAccount(ctx context.Context, account string, blobs []*blob.Blob, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	client.mtx.Lock()
+	defer client.mtx.Unlock()
+	if err := client.checkAccountLoaded(ctx, account); err != nil {
 		return nil, err
 	}
 
-	txBytes, _, err := s.signer.CreatePayForBlobs(account, blobs, opts...)
+	blobSizes := make([]uint32, len(blobs))
+	for i, blob := range blobs {
+		blobSizes[i] = uint32(len(blob.Data))
+	}
+
+	gasLimit := uint64(float64(types.DefaultEstimateGas(blobSizes)) * client.gasMultiplier)
+	fee := uint64(math.Ceil(appconsts.DefaultMinGasPrice * float64(gasLimit)))
+	// prepend calculated params, so it can be overwritten in case the user has specified it.
+	opts = append([]TxOption{SetGasLimit(gasLimit), SetFee(fee)}, opts...)
+
+	txBytes, _, err := client.signer.CreatePayForBlobs(account, blobs, opts...)
 	if err != nil {
 		return nil, err
 	}
 
-	return s.broadcastTx(ctx, txBytes, account)
+	return client.broadcastTx(ctx, txBytes, account)
 }
 
 // SubmitTx forms a transaction from the provided messages, signs it, and submits it to the chain. TxOptions
 // may be provided to set the fee and gas limit.
-func (s *TxClient) SubmitTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	resp, err := s.BroadcastTx(ctx, msgs, opts...)
+func (client *TxClient) SubmitTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	resp, err := client.BroadcastTx(ctx, msgs, opts...)
 	if err != nil {
 		return resp, err
 	}
 
-	return s.ConfirmTx(ctx, resp.TxHash)
+	return client.ConfirmTx(ctx, resp.TxHash)
 }
 
-func (s *TxClient) BroadcastTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	account, err := s.getAccountNameFromMsgs(msgs)
+func (client *TxClient) BroadcastTx(ctx context.Context, msgs []sdktypes.Msg, opts ...TxOption) (*sdktypes.TxResponse, error) {
+	client.mtx.Lock()
+	defer client.mtx.Unlock()
+	account, err := client.getAccountNameFromMsgs(msgs)
 	if err != nil {
 		return nil, err
 	}
 
-	if err := s.checkAccountLoaded(ctx, account); err != nil {
+	if err := client.checkAccountLoaded(ctx, account); err != nil {
 		return nil, err
 	}
 
-	tx, account, _, err := s.signer.SignTx(msgs, opts...)
+	txBuilder, err := client.signer.txBuilder(msgs, opts...)
 	if err != nil {
 		return nil, err
 	}
 
-	txBytes, err := s.signer.EncodeTx(tx)
+	estimated := false
+	for _, coin := range txBuilder.GetTx().GetFee() {
+		if coin.Denom == appconsts.BondDenom {
+			estimated = true
+			break
+		}
+	}
+
+	gasLimit := txBuilder.GetTx().GetGas()
+	if float32(gasLimit) < appconsts.DefaultMinGasPrice {
```

Comment:
Can you explain you're reasoning here? So if someone submits a transaction with a gas limit that is less than the default min gas price? Shouldn't it be if the gasLimit is 0?

Reply:
It covers the case when the user provides a gasLimit that is less than `DefaultMinGasPrice`, so the tx will be failed. I think it could be useful if at some point we decide to update MinGasPrice.